### ISSUE TITLE
List of software that should be installed on Linux or macOS

### ISSUE THREAD

**Issue by seisman** [seismo-learn/seismology101#90]
Software are divided into four groups:

1. **System software**: necessary for system management, or very low-level
2. **Developing software**: required for programming and compiling research codes
3. **Command Line Tools**
4. **Other software**: optional but recommended.

---

### 1. System Software

- macOS: Command Line Tools for Xcode + Homebrew
- Fedora: dnf
- Ubuntu: apt

### 2. Developing software

- C compiler: GCC, Clang
- Fortran compiler: GFortran
- Intel compilers
- Java, i.e., OpenJDK
- Python, via anaconda
- make
- git

### 3. Command Line Tools

- dos2unix, unix2dos
- wget
- gawk
- grep 
- find
- sed
- tar

### 4. Other software

| Category | macOS | Fedora | Ubuntu |
| --- | --- | --- | --- |
| Terminal | iTerm 2 | Terminator | Terminator |
| Editors | VScode + Typoro | VScode | VScode |
| Unzip | the-unarchiver | unar | unar |
| Google Earth | Yes | Yes | Yes |
| Browser | Google Chrome + Microsoft Edge | Google Chrome | Google Chrome |
| QuickLook | Yes | | |
| Office | | WPS office | WPS office |
| Virtual Machine | VMware Fusion | |

**Comment by core-man** [comment]
What's the difference between awk and gawk in Linux?

**Comment by seisman** [comment]
> What's the difference between awk and gawk in Linux?

I'm not sure. Perhaps awk is just a symlink of gawk in Linux.

**Comment by core-man** [comment]
> > What's the difference between awk and gawk in Linux?
> 
> I'm not sure. Perhaps awk is just a symlink of gawk in Linux.

Yes, you are right. `/usr/bin/awk -> gawk` in my CenOS 7.

**Comment by seisman** [comment]
macOS has its own `awk` in `/usr/bin/awk`. The gawk from Homebrew provides two commands: `awk` and `gawk`, and `awk` is a symlink of `gawk`. So aftering installing Homebrew, both `awk` and `gawk` are from Homebrew and they're the same.

**Comment by seisman** [comment]
macOS awk:
```
usage: /usr/bin/awk [-F fs] [-v var=value] [-f progfile | 'prog'] [file ...]
```

GNU awk:
```
sage: awk [POSIX or GNU style options] -f progfile [--] file ...
Usage: awk [POSIX or GNU style options] [--] 'program' file ...
POSIX options:		GNU long options: (standard)
	-f progfile		--file=progfile
	-F fs			--field-separator=fs
	-v var=val		--assign=var=val
Short options:		GNU long options: (extensions)
	-b			--characters-as-bytes
	-c			--traditional
	-C			--copyright
	-d[file]		--dump-variables[=file]
	-D[file]		--debug[=file]
...
```

**Comment by core-man** [comment]
> macOS awk:
> 
> ```
> usage: /usr/bin/awk [-F fs] [-v var=value] [-f progfile | 'prog'] [file ...]
> ```
> 
> GNU awk:
> 
> ```
> sage: awk [POSIX or GNU style options] -f progfile [--] file ...
> Usage: awk [POSIX or GNU style options] [--] 'program' file ...
> POSIX options:		GNU long options: (standard)
> 	-f progfile		--file=progfile
> 	-F fs			--field-separator=fs
> 	-v var=val		--assign=var=val
> Short options:		GNU long options: (extensions)
> 	-b			--characters-as-bytes
> 	-c			--traditional
> 	-C			--copyright
> 	-d[file]		--dump-variables[=file]
> 	-D[file]		--debug[=file]
> ...
> ```

Yes, we may add this note in masOS setup?

**Comment by seisman** [comment]
> Yes, we may add this note in masOS setup?

I'll open a separate issue for this.

**Comment by seisman** [comment]
It seems that we all agree about categorizing software into 4 groups. If yes, then we should have a consistent section structure for macOS, Fedora and Ubuntu (WSL is an exception).

1. 安装系统
2. 系统软件
3. 编程开发环境
4. 常用命令行工具
5. 日常软件

**Comment by core-man** [comment]
> It seems that we all agree about categorizing software into 4 groups. If yes, then we should have a consistent section structure for macOS, Fedora and Ubuntu (WSL is an exception).
> 
> 1. 安装系统
> 2. 系统软件
> 3. 编程开发环境
> 4. 常用命令行工具
> 5. 日常软件

Sound great~

**Comment by core-man** [comment]
> 4. 常用命令行工具

命令行工具?

**Comment by core-man** [comment]
Do we want to add Free Intel® Software Development Tools?
https://software.intel.com/content/www/us/en/develop/articles/free-intel-software-developer-tools.html

Intel Parallel Studio XE Cluster is now Intel® oneAPI Toolkits

**Comment by seisman** [comment]
> Do we want to add Free Intel® Software Development Tools?
> https://software.intel.com/content/www/us/en/develop/articles/free-intel-software-developer-tools.html
> 
> Intel Parallel Studio XE Cluster is now Intel® oneAPI Toolkits

We can, but perhaps we can add it after we publish the first version of guides for Ubuntu and Fedora.

**Comment by seisman** [comment]
What we may want to do before closing the issue:

- [ ] Remove X11
- [ ] macOS's default C compiler is clang. Explain how to use GCC 
- [ ] Add intel compilers
- [ ] Recommend [terminator](https://gnometerminator.blogspot.com/p/introduction.html) on Linux?
- [ ] Install Google Chrome on macOS?
- [ ] [GNOME SUSHUI](https://en.wikipedia.org/wiki/GNOME_sushi): Quicklook for Linux?
- [ ] VirtualBox for Linux?

**Comment by core-man** [comment]
> * [ ]  [GNOME SUSHUI](https://en.wikipedia.org/wiki/GNOME_sushi): Quicklook for Linux?

Yes to for all the other except this. Do anyone use quick look in Linux? I think most of us use CML.

**Comment by seisman** [comment]
> > * [ ]   [GNOME SUSHUI](https://en.wikipedia.org/wiki/GNOME_sushi): Quicklook for Linux?
> 
> Yes to me for all the other except this. Do anyone use quick look in Linux? I think most of us use CML.

OK. Removed from the list.

**Comment by core-man** [comment]
> * [ ]  Remove X11

We may first move X11 for linux, macos and windows to the appedix of software before we remove it.

**Comment by seisman** [comment]
> > * [ ]   Remove X11
> 
> We may first move X11 for linux, macos and windows to the appedix of software before we remove it.

In "software", the only package that needs X11 is CPS330. I think it's OK to put X11 in CPS330 now.

**Comment by core-man** [comment]
Okay, we may move it to the appendix when we have more software depending on x11.

**Comment by core-man** [comment]
> > > * [ ]    Remove X11
> > 
> > 
> > We may first move X11 for linux, macos and windows to the appedix of software before we remove it.
> 
> In "software", the only package that needs X11 is CPS330. I think it's OK to put X11 in CPS330 now.

I remember in WSL, we need x11 to run `taup` GUI.

**Comment by seisman** [comment]
If that's true, then we still need to have x11 installed, but it's slightly different, because running gui doesn't need header files.

**Comment by core-man** [comment]
> If that's true, then we still need to have x11 installed, but it's slightly different, because running gui doesn't need header files.

Yes, X server is needed.

**Comment by seisman** [comment]
> > If that's true, then we still need to have x11 installed, but it's slightly different, because running gui doesn't need header files.
> 
> Yes, X server is needed.

OK. We need to update x11 instructions, but it's unclear to me which packages are required

**Comment by core-man** [comment]
> > > If that's true, then we still need to have x11 installed, but it's slightly different, because running gui doesn't need header files.
> > 
> > 
> > Yes, X server is needed.
> 
> OK. We need to update x11 instructions, but it's unclear to me which packages are required

For windows, I only kown `taup` which is rarely used as we usually use command line tools like `taup_time`.

**Comment by seisman** [comment]
> > > > If that's true, then we still need to have x11 installed, but it's slightly different, because running gui doesn't need header files.
> > > 
> > > 
> > > Yes, X server is needed.
> > 
> > 
> > OK. We need to update x11 instructions, but it's unclear to me which packages are required
> 
> For windows, I only kown `taup` which is rarely used as we usually use command line tools like `taup_time`.

SAC also needs X11. I believe all GUI software on Linux need X11.

**Comment by seisman** [comment]
Here is what I understand:

1. For most Linux users, we don't need to mention X11 at all, because X11 server is always installed by default
2. For macOS, it's unclear to me if X11 is required for GUI software like TauP
3. For Windows WSL users, X11 server is not installed by default. That's the main trouble.

**Comment by core-man** [comment]
> > > > > If that's true, then we still need to have x11 installed, but it's slightly different, because running gui doesn't need header files.
> > > > 
> > > > 
> > > > Yes, X server is needed.
> > > 
> > > 
> > > OK. We need to update x11 instructions, but it's unclear to me which packages are required
> > 
> > 
> > For windows, I only kown `taup` which is rarely used as we usually use command line tools like `taup_time`.
> 
> SAC also needs X11. I believe all GUI software on Linux need X11.
yes, that is true. but for windows, it is a little different.

"WSL 本身不支持图形界面，需要在 Windows 中安装 X Server 来接收和显示 Linux 中的图形界面。"

**Comment by seisman** [comment]
> "WSL 本身不支持图形界面，需要在 Windows 中安装 X Server 来接收和显示 Linux 中的图形界面。"

Hmmm, so it means for Windows WSL, we don't need to install X11 in WSL, but we need to install VcXsrv on Windows?

**Comment by core-man** [comment]
> Here is what I understand:
> 
> 1. For most Linux users, we don't need to mention X11 at all, because X11 server is always installed by default
> 2. For macOS, it's unclear to me if X11 is required for GUI software like TauP
>
> 3. For Windows WSL users, X11 server is not installed by default. That's the main trouble.

Or we may move x11 from linux, macos and wsl to software appendix? and mention x11 in possible software which may need it on some os.

**Comment by core-man** [comment]
> > "WSL 本身不支持图形界面，需要在 Windows 中安装 X Server 来接收和显示 Linux 中的图形界面。"
> 
> Hmmm, so it means for Windows WSL, we don't need to install X11 in WSL, but we need to install VcXsrv on Windows?

maybe this is the reason.

**Comment by seisman** [comment]
I unstalled X11 on my macOS, and taup still works.

**Comment by core-man** [comment]
I would suggest we can remove x11 on linux ans X server on windows cause they may not  be used by most users.

Then add a section in software appendix.

**Comment by core-man** [comment]
> I unstalled X11 on my macOS, and taup still works.

I think macos support GUI while windows need X server to receive wsl output.

If we use a window gui in windows, I think there will be no problem.

**Comment by seisman** [comment]
OK to remove X11 from the setup guides. Please review https://github.com/seismo-learn/seismology101/pull/101

**Comment by zhaozhiyuan1989** [comment]
@seisman @zhaozhiyuan1989 我在WSL2版本的Ubuntu上按照在一般Linux上的步骤安装了TauP、SAC、MATLAB等，都能在 Windows 上利用 X Server 正确打开GUI

**Comment by seisman** [comment]
@zhaozhiyuan1989 How did you install SAC? Did you use the binary package or compile from source codes?

**Comment by zhaozhiyuan1989** [comment]
> @zhaozhiyuan1989 How did you install SAC? Did you use the binary package or compile from source codes?

From source codes.

**Comment by core-man** [comment]
> @seisman @zhaozhiyuan1989 我在WSL2版本的Ubuntu上按照在一般Linux上的步骤安装了TauP、SAC、MATLAB等，都能在 Windows 上利用 X Server 正确打开GUI

can you run matlab on windows if you directly install it on windows instead of wsl?

**Comment by seisman** [comment]
> > @zhaozhiyuan1989 How did you install SAC? Did you use the binary package or compile from source codes?
> 
> From source codes.

When you compile SAC source codes, you need to install X11, right? If you install TauP after compiling SAC, then X11 is already installed, so it's still unclear to me if TauP depends on X11.

**Comment by core-man** [comment]
> > > @zhaozhiyuan1989 How did you install SAC? Did you use the binary package or compile from source codes?
> > 
> > 
> > From source codes.
> 
> When you compile SAC source codes, you need to install X11, right? If you install TauP after compiling SAC, then X11 is already installed, so it's still unclear to me if TauP depends on X11.

It could be that `taup` is a GUI, and wsl do not support GUI so we need x server for taup using WSL.

**Comment by seisman** [comment]
We won't have a conclusion unless we can test it. I think we can just ignore it now.

**Comment by core-man** [comment]
On Windows, we do not install X. we only install X server.

**Comment by zhaozhiyuan1989** [comment]
> > > @zhaozhiyuan1989 How did you install SAC? Did you use the binary package or compile from source codes?
> > 
> > 
> > From source codes.
> 
> When you compile SAC source codes, you need to install X11, right? If you install TauP after compiling SAC, then X11 is already installed, so it's still unclear to me if TauP depends on X11.

我先安装的 TauP，只需要安装 Java 就可以转发图形界面

**Comment by zhaozhiyuan1989** [comment]
也许等我有时间了可以在安装一个新的WSL2来测试一下。你们可以告诉我都需要测试什么。

**Comment by seisman** [comment]
> > > > @zhaozhiyuan1989 How did you install SAC? Did you use the binary package or compile from source codes?
> > > 
> > > 
> > > From source codes.
> > 
> > 
> > When you compile SAC source codes, you need to install X11, right? If you install TauP after compiling SAC, then X11 is already installed, so it's still unclear to me if TauP depends on X11.
> 
> 我先安装的 TauP，只需要安装 Java 就可以转发图形界面

Good to know.

**Comment by core-man** [comment]
> > > > @zhaozhiyuan1989 How did you install SAC? Did you use the binary package or compile from source codes?
> > > 
> > > 
> > > From source codes.
> > 
> > 
> > When you compile SAC source codes, you need to install X11, right? If you install TauP after compiling SAC, then X11 is already installed, so it's still unclear to me if TauP depends on X11.
> 
> 我先安装的 TauP，只需要安装 Java 就可以转发图形界面

Did you install x server before using taup.

**Comment by zhaozhiyuan1989** [comment]
我全新安装了一个 WSL2 Ubuntu20.04LTS ，安装 SAC 二进制包，在没有安装任何依赖的情况下运行 sac 会报错
`sac: error while loading shared libraries: libncurses.so.5: cannot open shared object file: No such file or directory`

`sudo apt install libncurses` 后即可正常运行 SAC 并显示图像

**Comment by core-man** [comment]
Maybe WSL has developed to a better version.

**Comment by core-man** [comment]
> 我全新安装了一个 WSL2 Ubuntu20.04LTS ，安装 SAC 二进制包，
> `libc6 libsm6 libice6 libxpm4 libx11-6 zlib1g` 均已默认安装
> 安装 `sudo apt install libncurses5` 后即可正常运行 SAC 并显示图像
> 
> `libx11-6 libx11-dev` 好像默认安装了？

try `apt list --installed | grep libx11` ?

**Comment by zhaozhiyuan1989** [comment]
> apt list --installed | grep libx11

初始安装完成后运行结果：
```
zhao@ZZY-DELL-G7:/mnt/c/Users/shetu$ apt list --installed | grep libx11

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

libx11-6/focal,now 2:1.6.9-2ubuntu1 amd64 [installed,automatic]
libx11-data/focal,now 2:1.6.9-2ubuntu1 all [installed,automatic]
libx11-xcb1/focal,now 2:1.6.9-2ubuntu1 amd64 [installed]
```

**Comment by core-man** [comment]
![image](https://user-images.githubusercontent.com/50591376/105452908-1fac7380-5cba-11eb-96c1-3214c103039e.png)

I don't the difference between those libraries.

**Comment by core-man** [comment]
> 安装 `sudo apt install libncurses5` 后即可正常运行 SAC 并显示图像

What if you don't install `libncurses5`?

**Comment by zhaozhiyuan1989** [comment]
> > 安装 `sudo apt install libncurses5` 后即可正常运行 SAC 并显示图像
> 
> What if you don't install `libncurses5`?

启用 sac 报错 `sac: error while loading shared libraries: libncurses.so.5: cannot open shared object file: No such file or directory`

**Comment by seisman** [comment]
The biggest difference is: libx11-6 only provides libraries, while libx11-dev provides both header files and libraries (libx11-dev depends on libx11-6).

When you run SAC, you need libraries, so libx11-6 must be installed. 

When you want to build SAC from source codes, you need both libraries and headers, so you have to install libx11-dev.

That's also why the libx11-dev package has the `-dev` suffix, because you only need it when you're developing something (i.e., compileing a package from source).

**Comment by seisman** [comment]
Here are my second thoughts of these to-dos:

> * [ ]  macOS's default C compiler is clang. Explain how to use GCC

I've been using the Apple Clang compilers for a long time, and I don't have any issues when compiling C codes. Perhpas it's unnecessary to mention GCC anymore, as it may cause more incompatibilities and I have no experience using it under macOS.

> * [ ]  Add intel compilers

I have no need so won't work on it recently.

> * [ ]  Recommend [terminator](https://gnometerminator.blogspot.com/p/introduction.html) on Linux

Sounds good to me, but I haven't used it for a long time.

> * [ ]  VirtualBox for Linux

As I understand it, installing Virtualbox also means add more things to Linux kernels, which may make Linux unstable and 
may cause more troubles when upgrading the Linux OS. I'm a little hesitant to add it.

**Comment by core-man** [comment]
> Here are my second thoughts of these to-dos:
> 
> > * [ ]   macOS's default C compiler is clang. Explain how to use GCC
> 
> I've been using the Apple Clang compilers for a long time, and I don't have any issues when compiling C codes. Perhpas it's unnecessary to mention GCC anymore, as it may cause more incompatibilities and I have no experience using it under macOS.

If there is no issue using Apple Clang, we may just add a note so the new babies know such a thing while they don't need to explore it right now.

> > * [ ]   Add intel compilers
> 
> I have no need so won't work on it recently.

I'll work on it since I'm using and must use it now on centos 7.

> > * [ ]   Recommend [terminator](https://gnometerminator.blogspot.com/p/introduction.html) on Linux
> 
> Sounds good to me, but I haven't used it for a long time. It's unclear to me if it's still one of the best terminals.

I will work on it and find if there is a better one.

> > * [ ]   VirtualBox for Linux
> 
> As I understand it, installing Virtualbox also means add more things to Linux kernels, which may make Linux unstable and
> may cause more troubles when upgrading the Linux OS. I'm a little hesitant to add it.

For MacOS, we may use virtual machine sometimes.
For Windows, we can use WLS now.
For Linux, we may sometimes have to use Windows. But I guess everyone has a PC with either Windows or macOS. They can use Windows directly or use a virtual machines for windows in macOS.

So I am okay to ignore this topic until we really need it someday.

**Comment by core-man** [comment]
> > * [ ]   Recommend [terminator](https://gnometerminator.blogspot.com/p/introduction.html) on Linux
> 
> Sounds good to me, but I haven't used it for a long time. It's unclear to me if it's still one of the best terminals.

It seems that Terminator is still one of the most popular terminals: https://fossbytes.com/best-linux-terminal-emulators/

**Comment by core-man** [comment]
Do we want to mention openMP and MPI?
https://blog.seisman.info/centos7-setup-2/#%E5%B9%B6%E8%A1%8C%E8%AE%A1%E7%AE%97

**Comment by seisman** [comment]
> Do we want to mention MPI?
> https://blog.seisman.info/centos7-setup-2/#%E5%B9%B6%E8%A1%8C%E8%AE%A1%E7%AE%97
> 
> My suggestion is that not everyone will use MPI and we are not MPI experts, so we may not include it here.
> 
> As we discussed before (#30), we can find someone to write a section about parallel computation for C and Fortran (openMP and MPI) and Python (multiple threads and processors).

It's low priority to me.

**Comment by core-man** [comment]
> > Do we want to mention MPI?
> > https://blog.seisman.info/centos7-setup-2/#%E5%B9%B6%E8%A1%8C%E8%AE%A1%E7%AE%97
> > My suggestion is that not everyone will use MPI and we are not MPI experts, so we may not include it here.
> > As we discussed before (#30), we can find someone to write a section about parallel computation for C and Fortran (openMP and MPI) and Python (multiple threads and processors).
> 
> It's low priority to me.

Me, too. Let's keep it on todo list and hope a nice guy will come up.

**Comment by core-man** [comment]
> What we may want to do before closing the issue:
> 
> * [x]  Remove X11
> * [x]  macOS's default C compiler is clang. Explain how to use GCC
> * [x]  Add intel compilers
> * [x]  Recommend [terminator](https://gnome-terminator.org/) or a better one on Linux
> * [x]  Install Google Chrome on macOS
> * [ ]  ~[GNOME SUSHUI](https://en.wikipedia.org/wiki/GNOME_sushi): Quicklook for Linux?~
> * [ ]  ~VirtualBox for Linux~

Done~
